### ISSUE TITLE
Input delay when authenticated (Google) on Firefox

### ISSUE THREAD

**Issue by crouton111** [raithos/xwing#707]
### Expected Behavior

Normal (minimal) input delay while authenticated (Google) on Firefox.

### Actual Behavior

When selecting a ship, pilot, or any other item from a drop-down, there is an input delay of about 4-5 seconds. Hover effects are also delayed. This behavior is not observed on Firefox when I'm not authenticated. Also, it's not observed on Chrome in either authentication status.

### Language

English

**Comment by raithos** [comment]
yeah, there seems to be a couple of seconds delay when you load in the page logged in, where the log in process stalls loading of the page. once you've logged in the input delay is gone. Not sure whats causing it, or why it only seems to happen on firefox

**Comment by crouton111** [comment]
In my case the input delay remains after page load. Once I'm logged in the delay is persistent. So if I'm making a rebel squad, there's a delay while choosing a ship chassis. Then a delay when selecting a pilot, another with selecting a torpedo upgrade, and so on. I'm running two browser extensions, but I disabled them both and there was no change. The only other thing I can think of that's environmental is that I always use Private mode in Firefox.

**Comment by crouton111** [comment]
I grabbed a video of it in action if it helps. Note, that I'm clicking the left mouse button to select something as soon as I'm hovering over it (it's otherwise hard to tell when the mouse click occurs from the video).

https://user-images.githubusercontent.com/39347503/104113052-cdd01a80-52c3-11eb-88a9-c3da834db3b1.mp4

**Comment by raithos** [comment]
woah that's actually crazy. Does this happen when you log in with a different account as well? I haven't been able to replicate the issue locally. Also, what version of firefox are you using?

**Comment by crouton111** [comment]
Yeah, really weird. I just tried it with a Discord login and the same thing is happening, which is interesting since I thought it might have had something to do with the collection querying, but I didn't enter my collection with the Discord login, whereas my Google login does have my collection.
The version of Firefox is 84.0.2 (64-bit) on Win10.

**Comment by raithos** [comment]
so i made some changes to how login triggers. could you check to see that the issue is still there? it probably is but there's an off chance it got cleared out with some of the optimization ive done

**Comment by crouton111** [comment]
Cleared my cache and the issue still appears to be there. I checked from another machine as well with the same results. I'm currently on v86.0 of Firefox.

**Comment by raithos** [comment]
does this happen to you on chrome or edge as well?

**Comment by raithos** [comment]
ok im able to reproduce it! not sure how to fix it!

**Comment by crouton111** [comment]
I haven't tried Edge, but Chrome does not exhibit the delay.

**Comment by Patschke** [comment]
First of all: It does not take that long by any means for me when selecting a ship. It does however take ~0.5s, which can of course be sped up - so I'll just assume, that the 0.5s for me have the same origin as the ~10s for you. I've tried to do some profiling on what happens during this time: 

![grafik](https://user-images.githubusercontent.com/17853362/111902061-bf2c7f00-8a3b-11eb-8ddd-cb7f71f8b18a.png)

The mouse-up event is of course the important part of the click. (The later DOM events are some hovering over stuff beeing also part of the caputred time frame)

Here is a function call tree for this time frame: 
![grafik](https://user-images.githubusercontent.com/17853362/111902135-22b6ac80-8a3c-11eb-8698-85ed6d64b8ba.png)

Half of the time is spend collecting garbage, so we could try to be a bit more carfeul with our resource management. 

![grafik](https://user-images.githubusercontent.com/17853362/111902236-be481d00-8a3c-11eb-8ef7-e180d4538d0f.png)

During the first half (where actually stuff other than gc happens) ~2/3 of the time seem to be spend on usefull things (e.g. creating the actual slots). However, the last ~1/3 of this time (not visible above) seems to be a few hundred (or thousand) issued from checks against the collection:

![grafik](https://user-images.githubusercontent.com/17853362/111902329-580fca00-8a3d-11eb-8717-36183f673203.png)

I can imagine, that these calls sum up to a longer time, if the collection is larger or there is any other reason why it's handling is slow. 

I have trouble to debug this properly, as I can't log in while using my local YASB build - due to Cross-Origin restrictions on https://yasb-backend.herokuapp.com. Not sure if I have to set up my own backend for testing, or if there is another way to circuvent this...

**Comment by raithos** [comment]
So this is the conclusion I'm running up against as well. From my testing last night as well I've come to the same conclusion: hundreds of checks against the collection. It looks like when we add a new upgrade it checks every slots upgrades against the entire library of upgrades in collection whether you have something or not. But I'm not sure of a more efficient way to rewrite this atm

**Comment by Patschke** [comment]
There should be two options conceptually: We can either keep track of the collection more rigourously (currently `@collection.use` is only ever called in `isSquadPossibleWithCollection`, we could instead do so on adding stuff. Of course, this requires to add an unuse method as well, since we currently only use a reset). The other option is to prevent pointless calls more efficiently, either by using `@currenty_squad.dirty` or toggling `collectioncheck`?

**Comment by crouton111** [comment]
If it's any help here are the contents of my Collection from the Inventory tab:

Ship

    A/SF-01 B-Wing - 2
    Aggressor Assault Fighter - 3
    Alpha-class Star Wing - 2
    Attack Shuttle - 1
    BTL-A4 Y-Wing - 3
    Customized YT-1300 Light Freighter - 1
    E-Wing - 1
    Escape Craft - 1
    Fang Fighter - 2
    Firespray-class Patrol Craft - 3
    G-1A Starfighter - 1
    HWK-290 Light Freighter - 1
    JumpMaster 5000 - 1
    Kihraxz Fighter - 2
    Lambda-class T-4a Shuttle - 2
    Lancer-class Pursuit Craft - 1
    M12-L Kimogila Fighter - 2
    M3-A Interceptor - 1
    Mining Guild TIE Fighter - 1
    Modified YT-1300 Light Freighter - 1
    Quadrijet Transfer Spacetug - 4
    RZ-1 A-Wing - 1
    Scurrg H-6 Bomber - 1
    Sheathipede-class Shuttle - 1
    StarViper-class Attack Platform - 2
    T-65 X-Wing - 6
    TIE Advanced v1 - 1
    TIE Advanced x1 - 2
    TIE/D Defender - 1
    TIE Interceptor - 2
    TIE Reaper - 1
    TIE/ca Punisher - 1
    TIE/ln Fighter - 6
    TIE/ph Phantom - 1
    TIE/sa Bomber - 4
    TIE/sk Striker - 1
    UT-60D U-Wing - 1
    VCX-100 Light Freighter - 1
    YT-2400 Light Freighter - 1
    YV-666 Light Freighter - 1
    Z-95-AF4 Headhunter - 1

Pilot

    4-LOM - 2
    AP-5 - 1
    Academy Pilot - 10
    Ahhav - 1
    Airen Cracken - 1
    Alpha Squadron Pilot - 2
    Arvel Crynyd - 1
    Asajj Ventress - 2
    Autopilot Drone - 1
    Bandit Squadron Pilot - 3
    Baron of the Empire - 3
    Benthic Two Tubes - 1
    Biggs Darklighter - 3
    Binayre Pirate - 6
    Black Squadron Ace - 10
    Black Squadron Scout - 4
    Black Sun Ace - 6
    Black Sun Assassin - 4
    Black Sun Enforcer - 4
    Black Sun Soldier - 6
    Blade Squadron Veteran - 3
    Blue Squadron Escort - 6
    Blue Squadron Pilot - 3
    Blue Squadron Scout - 2
    Boba Fett - 3
    Bodhi Rook - 1
    Bossk - 2
    Bounty Hunter - 5
    Braylen Stramm - 2
    Captain Feroph - 1
    Captain Jonus - 1
    Captain Jostero - 2
    Captain Kagi - 1
    Captain Nym - 2
    Captain Oicunn - 1
    Captain Rex - 1
    Captain Seevor - 1
    Cartel Executioner - 4
    Cartel Marauder - 6
    Cartel Spacer - 8
    Cassian Andor - 1
    Cavern Angels Zealot - 3
    Chewbacca - 1
    “Chopper” - 1
    Colonel Jendon - 1
    Colonel Vessery - 1
    Constable Zuvio - 2
    Contracted Scout - 2
    Corran Horn - 1
    “Countdown” - 2
    Countess Ryad - 1
    Crymorah Goon - 4
    Cutlass Squadron Pilot - 2
    Dace Bonearm - 2
    Dalan Oberos - 2
    Dalan Oberos - 2
    Darth Vader - 2
    Dash Rendar - 1
    “Deathfire” - 1
    “Deathrain” - 1
    Del Meeko - 3
    Delta Squadron Pilot - 2
    Dengar - 2
    “Double Edge” - 1
    Drea Renthal - 2
    “Duchess” - 2
    “Dutch” Vander - 3
    “Echo” - 1
    Edrio Two Tubes - 1
    Emon Azzameen - 3
    Esege Tuketu - 1
    Evaan Verlaine - 3
    Ezra Bridger - 1
    Ezra Bridger - 1
    Ezra Bridger - 1
    Fenn Rau - 4
    Fenn Rau - 1
    Foreman Proach - 1
    Freighter Captain - 1
    Gamma Squadron Ace - 3
    Gand Findsman - 4
    Garven Dreis - 3
    Garven Dreis - 1
    Gavin Darklighter - 1
    Genesis Red - 2
    Gideon Hask - 3
    Gold Squadron Veteran - 4
    Grand Inquisitor - 1
    Gray Squadron Bomber - 4
    Graz - 2
    Green Squadron Pilot - 2
    Guri - 2
    Han Solo - 1
    Han Solo - 1
    Heff Tobber - 1
    Hera Syndulla - 1
    Hera Syndulla - 1
    Hired Gun - 4
    Horton Salm - 3
    “Howlrunner” - 3
    IG-88A - 2
    IG-88B - 2
    IG-88C - 2
    IG-88D - 2
    Ibtisam - 1
    Iden Versio - 2
    Imdaar Test Pilot - 2
    Inaldra - 2
    Inquisitor - 3
    Jake Farrell - 1
    Jakku Gunrunner - 6
    Jan Ors - 1
    Jek Porkins - 2
    Joy Rekkoff - 4
    Kaa’to Leeachos - 2
    Kad Solus - 4
    Kanan Jarrus - 1
    Kashyyyk Defender - 2
    Kath Scarlet - 3
    Kavil - 2
    Ketsu Onyo - 2
    Knave Squadron Escort - 2
    Koshka Frost - 3
    Krassis Trelix - 3
    Kullbee Sperado - 1
    Kyle Katarn - 1
    L3-37 - 1
    L3-37 - 1
    Laetin A’shera - 2
    Lando Calrissian - 1
    Lando Calrissian - 1
    Lando Calrissian - 1
    Latts Razzi - 2
    “Leebo” - 1
    Leevan Tenza - 1
    Lieutenant Blount - 1
    Lieutenant Karsabi - 1
    Lieutenant Kestal - 1
    Lieutenant Sai - 1
    Lok Revenant - 4
    Lothal Rebel - 1
    Lowhhrick - 1
    Luke Skywalker - 2
    Maarek Stele - 2
    Magva Yarro - 1
    Major Rhymer - 1
    Major Vermeil - 1
    Major Vynder - 1
    Manaroo - 2
    “Mauler” Mithel - 3
    Mining Guild Sentry - 1
    Mining Guild Surveyor - 1
    Miranda Doni - 1
    Moralo Eval - 2
    Nashtah Pup - 2
    N’dru Suhlak - 2
    “Night Beast” - 2
    Norra Wexley - 3
    Norra Wexley - 1
    Nu Squadron Pilot - 2
    Obsidian Squadron Pilot - 10
    Old Teroch - 4
    Omicron Group Pilot - 1
    Onyx Squadron Ace - 2
    Onyx Squadron Scout - 2
    Outer Rim Pioneer - 1
    Outer Rim Smuggler - 1
    Overseer Yushyn - 1
    Palob Godalhi - 2
    Partisan Renegade - 1
    Patrol Leader - 2
    Phoenix Squadron Pilot - 2
    Planetary Sentinel - 4
    Prince Xizor - 2
    “Pure Sabacc” - 2
    Quinn Jast - 2
    Rear Admiral Chiraneau - 1
    Rebel Scout - 2
    Red Squadron Veteran - 6
    “Redline” - 1
    Rexler Brath - 1
    Rho Squadron Pilot - 2
    Roark Garnet - 1
    Rogue Squadron Escort - 2
    Saber Squadron Ace - 2
    Sabine Wren - 1
    Sabine Wren - 1
    Sabine Wren - 2
    Sarco Plank - 2
    Saw Gerrera - 1
    Scarif Base Pilot - 1
    Scimitar Squadron Pilot - 3
    “Scourge” Skutu - 3
    Serissu - 2
    Seventh Sister - 1
    Seyn Marana - 3
    Shadowport Hunter - 2
    Shara Bey - 1
    Sienar Specialist - 2
    Sigma Squadron Ace - 2
    Skull Squadron Pilot - 6
    Sol Sixxa - 2
    Soontir Fel - 1
    Spice Runner - 6
    Storm Squadron Ace - 3
    Sunny Bounder - 2
    Tala Squadron Pilot - 3
    Talonbane Cobra - 2
    Tansarii Point Veteran - 8
    Tel Trevura - 2
    Tempest Squadron Pilot - 3
    Ten Numb - 2
    Thane Kyrell - 3
    Tomax Bren - 1
    Torani Kulda - 2
    Torkil Mux - 2
    Trandoshan Slaver - 2
    Turr Phennir - 1
    Unkar Plutt - 2
    Valen Rudor - 2
    Ved Foslo - 2
    Viktor Hel - 2
    “Vizier” - 1
    “Wampa” - 3
    Warden Squadron Pilot - 2
    Wedge Antilles - 3
    “Whisper” - 1
    Wild Space Fringer - 1
    Wullffwarro - 1
    Zealous Recruit - 8
    “Zeb” Orrelios - 1
    “Zeb” Orrelios - 1
    “Zeb” Orrelios - 1
    Zertik Strom - 2
    Zuckuss - 2

Upgrade

    0-0-0 - 2
    4-LOM - 2
    Ablative Plating - 8
    Admiral Sloane - 1
    Adv. Proton Torpedoes - 10
    Advanced SLAM - 4
    Advanced Sensors - 9
    Afterburners - 5
    Agent Kallus - 1
    Agile Gunner - 1
    Andrasta - 3
    BT-1 - 2
    Barrage Rockets - 3
    Baze Malbus - 1
    Bistan - 1
    Boba Fett - 3
    Bomblet Generator - 4
    Bossk - 2
    C-3PO - 1
    Cad Bane - 2
    Cassian Andor - 1
    Chewbacca - 1
    Chewbacca - 1
    “Chopper” - 1
    “Chopper” - 1
    Ciena Ree - 1
    Cikatro Vizago - 2
    Cloaking Device - 3
    Cluster Missiles - 10
    Collision Detector - 8
    Composure - 1
    Concussion Missiles - 10
    Conner Nets - 10
    Contraband Cybernetics - 6
    Crack Shot - 13
    Daredevil - 10
    Darth Vader - 1
    Dauntless - 1
    Deadman’s Switch - 10
    Death Troopers - 1
    Debris Gambit - 8
    Dengar - 2
    Director Krennic - 1
    Dorsal Turret - 8
    Electronic Baffle - 9
    Elusive - 11
    Emperor Palpatine - 1
    Engine Upgrade - 6
    Expert Handling - 10
    Ezra Bridger - 1
    Fearless - 8
    Feedback Array - 8
    Fifth Brother - 1
    Fire-Control System - 10
    Foresight - 3
    Freelance Slicer - 8
    GNK “Gonk” Droid - 8
    “Genius” - 2
    Ghost - 1
    Grand Inquisitor - 1
    Grand Moff Tarkin - 1
    Greedo - 2
    Han Solo - 1
    Han Solo - 1
    Havoc - 2
    Heavy Laser Cannon - 10
    Heightened Perception - 5
    Hera Syndulla - 1
    Homing Missiles - 9
    Hotshot Gunner - 8
    Hound’s Tooth - 2
    Hull Upgrade - 11
    IG-2000 - 4
    IG-88D - 2
    ISB Slicer - 2
    Inertial Dampeners - 7
    Informant - 4
    Instinctive Aim - 2
    Intimidation - 13
    Ion Cannon - 9
    Ion Cannon Turret - 10
    Ion Missiles - 9
    Ion Torpedoes - 13
    Jabba the Hutt - 2
    Jamming Beam - 9
    Juke - 13
    Jyn Erso - 1
    Kanan Jarrus - 1
    Ketsu Onyo - 2
    L3-37 - 1
    Lando Calrissian - 1
    Lando Calrissian - 1
    Lando’s Millennium Falcon - 1
    Latts Razzi - 2
    Leia Organa - 1
    Lone Wolf - 5
    Luke Skywalker - 1
    Magva Yarro - 1
    Marauder - 3
    Marksmanship - 10
    Maul - 2
    Millennium Falcon - 1
    Minister Tua - 1
    Mist Hunter - 2
    Moff Jerjerrod - 1
    Moldy Crow - 1
    Munitions Failsafe - 8
    Nien Nunb - 1
    Novice Technician - 8
    Os-1 Arsenal Loadout - 3
    Outmaneuver - 11
    Outrider - 1
    Perceptive Copilot - 9
    Phantom - 1
    Pivot Wing - 3
    Predator - 11
    Proton Bombs - 12
    Proton Rockets - 9
    Proton Torpedoes - 12
    Proximity Mines - 10
    Punishing One - 2
    Qi’ra - 1
    R2 Astromech - 10
    R2-D2 - 1
    R2-D2 - 2
    R3 Astromech - 9
    R4 Astromech - 9
    R5 Astromech - 10
    R5-D8 - 2
    R5-P8 - 2
    R5-TK - 2
    Rigged Cargo Chute - 7
    Ruthless - 4
    ST-321 - 1
    Sabine Wren - 1
    Saturation Salvo - 8
    Saw Gerrera - 1
    Seasoned Navigator - 8
    Seismic Charges - 12
    Selfless - 5
    Sense - 4
    Servomotor S-Foils - 7
    Seventh Sister - 1
    Shadow Caster - 2
    Shield Upgrade - 10
    Skilled Bombardier - 9
    Slave I - 3
    Snap Shot - 3
    Squad Leader - 6
    Static Discharge Vanes - 9
    Stealth Device - 10
    Supernatural Reflexes - 5
    Swarm Tactics - 11
    Tactical Officer - 9
    Tactical Scrambler - 9
    Tobias Beckett - 1
    Tractor Beam - 8
    Trajectory Simulator - 6
    Trick Shot - 13
    Unkar Plutt - 2
    Veteran Tail Gunner - 7
    Veteran Turret Gunner - 10
    Virago - 2
    Xg-1 Assault Configuration - 3
    “Zeb” Orrelios - 1
    Zuckuss - 2

Summary

    Ship - 72 (41 different)
    Pilot - 519 (235 different)
    Upgrade - 804 (166 different)

**Comment by raithos** [comment]
@crouton111 i cleaned up your inventory list because it makes reading the convo more difficult. This bug happens on empty collections so its not specifically an inventory issue

**Comment by crouton111** [comment]
No worries.

**Comment by XPav** [comment]
This is also tied to "Address Autofill".   Turn this off and it's much zippier (although still a little poky compared to Edge hahah I meant Chrome).


![image](https://user-images.githubusercontent.com/10382630/112241057-c18f0480-8c06-11eb-8986-0276464d454c.png)

**Comment by raithos** [comment]
@crouton111  please try this and see if it fixes it. it did fix it when I did it.

**Comment by crouton111** [comment]
@raithos I unchecked "Autofill addresses", cleared my cache, and the issue is still there.

**Comment by raithos** [comment]
So now because of the autofill I cannot reproduce the bug. So i'll have to think about this more. It's possible its related to autofill in that its triggering some sort of weird interaction with firefox, but I have to think about it more.

**Comment by XPav** [comment]
Issue #261 was a previous issue where we went back and forth and it got way better, FWIW.

**Comment by raithos** [comment]
Hey, I changed a bunch of other things, and I'm wondering if firefox performance has improved? Right now when i load it it loads exactly the same as chrome

**Comment by XPav** [comment]
No performance improvement for me on Firefox 87.0 when logged in.

FWIW I'm likely switching to Edge as my daily browser (because I hates the Google?) and Edge works just as zippy as Chrome.

**Comment by crouton111** [comment]
No improvement for me either.

**Comment by raithos** [comment]
so im on 88.0.1 and still not experiencing any issues with lag. are the issues still there?

**Comment by crouton111** [comment]
I'm also on 88.0.1 and the issue is still there for me.

**Comment by raithos** [comment]
hey guys can you also check that your search and form history is also turned off? apparently firefox saves your preference even if you have autofill turned off

https://support.mozilla.org/en-US/questions/1265289

**Comment by crouton111** [comment]
My setting is
Firefox will: "Never remember history"
...which precludes those three checkboxes.

**Comment by raithos** [comment]
I hate doing this every so often, but I'm still unable to reproduce the error on FF 89.0.2. Is this still a problem for you?

**Comment by XPav** [comment]
I switched to Edge, so I don't have any problems anymore.

**Comment by raithos** [comment]
I'm going to close this as there is nothing I can do to improve this until we get more information

**Comment by asitnick** [comment]
Sorry for commenting on a closed case, but I wanted to mention that I've experienced this issue forever. For me, it didn't matter what service I used for logging in, I would always get a long delay whenever accessing or selecting something from one of the droplists. But I started using a Password Manager, so today tried I disabling "Autofill Addresses" (since the password manager handles that) for the first time. The rest of my settings are defaults. I reloaded YASB, and everything behaves as it does in Chrome or Edge. I'm using Twitter as my service to save lists.
